feat: doc/extension charts (#347)
Avaiga/taipy-doc: docs/userman/gui/extension/chart.md

# Chart

> In this section, we’ll dive into creating and integrating charts using [`Plotly`](https://plotly.com/graphing-libraries/) in your extension library.

Reviewer: ```suggestion
In this section, we’ll dive into creating and integrating charts using [`Plotly`](https://plotly.com/graphing-libraries/) in our extension library.
```

Charts are invaluable for visualizing data, offering clear and interactive insights that enhance the user experience.
We’ll walk through the steps to implement different types of charts, configure their properties,
and seamlessly embed them into your application.

## Prerequisites

Before you start creating charts, ensure you have the following prerequisites:

> - **Install the necessary libraries:** To use Plotly.js with React, you’ll need to install the react-plotly.js library.

Reviewer: ```suggestion
- **Install the necessary libraries:** To use [Plotly.js with React](https://plotly.com/javascript/react/), you’ll need to install the [`react-plotly.js`](https://github.com/plotly/react-plotly.js) library.
```

  ```bash
  $ npm install react-plotly.js plotly.js
  ```

> !!!Note

Reviewer: ```suggestion
!!! note "Using TypeScript"
```

    When using TypeScript, you’ll need to install the type definitions for Plotly.js as well.

    ```bash
    $ npm install --dev @types/react-plotly.js
    ```

> ## Declaring the element {data-source="gui:doc/extension/example_library/example_library.py#L69"}

Reviewer: ```suggestion
## Declaring the element {data-source="gui/extension/example_library/example_library.py#L69"}
```
Yeah this has changed for the better.

> To create a chart element, you need to define the element in your extension library.

Reviewer: ```suggestion
To create an element that holds a chart, you need to define it in the extension library:
```

```python title="example_library.py"
from taipy.gui.extension import Element, ElementLibrary, ElementProperty, PropertyType

class ExampleLibrary(ElementLibrary):
    def __init__(self) -> None:
        # Initialize the set of visual elements for this extension library
        self.elements = {
            "dashboard": Element(
                "data",
                {
                    "data": ElementProperty(PropertyType.dict),
                    "layout": ElementProperty(PropertyType.dict),
                },
            )
        }
```

The detailed explanation of the code is as follows:

- The *dashboard* element includes two properties: *data* and *layout*.

> - The *data* property has the type `PropertyType.dict`, meaning it holds a dictionary of data.

Reviewer: Can PropertyType.dict be a crossref?

- The *layout* property has the type `PropertyType.dict`, meaning it holds a dictionary of layout properties.

> ## Creating the React component {data-source="gui:doc/extension/example_library/front-end/src/Dashboard.tsx"}

Reviewer: ```suggestion
## Creating the React component {data-source="gui/extension/example_library/front-end/src/Dashboard.tsx"}
```

The React component for the *dashboard* element is defined as follows:

```tsx title="Dashboard.tsx"
import React, { useMemo } from "react";
import { useDynamicJsonProperty, } from "taipy-gui";

import Plot from "react-plotly.js";
import { Data, Layout } from "plotly.js";

interface DashboardProps {
    data?: string;
    defaultData?: string;
    layout?: string;
    defaultLayout?: string;
}

const Dashboard = (props: DashboardProps) => {
    const value = useDynamicJsonProperty(props.data, props.defaultData || "", {} as Partial<Data>);
    const dashboardLayout = useDynamicJsonProperty(props.layout, props.defaultLayout || "", {} as Partial<Layout>);

>     const data = useMemo(() => {

Reviewer: A short comment explaining why we're using Memos would be nice

        if (Array.isArray(value)) {
            return value as Data[];
        }
        return [] as Data[];
    }, [value]);

    const baseLayout = useMemo(() => {
        const layout = {
            ...dashboardLayout,
        };
        return layout as Partial<Layout>;
    }, [dashboardLayout]);

    return (

>         <div>

Reviewer: Why do we need the div?

            <Plot data={data} layout={baseLayout} />
        </div>
    );
};

export default Dashboard;
```

The detailed explanation of the code is as follows:

- The **Dashboard** component accepts two props: *data* and *layout*.
- The [`useDynamicJsonProperty()`](../../../refmans/reference_guiext/functions/useDynamicJsonProperty.md) hook is used to handle dynamic properties. It accepts the following arguments:

    - *data*: The dynamic property that holds the data for the chart.

>     - *defaultData*: The default value for the data property.

Reviewer: ```suggestion
    - *defaultData*: The default value for the *data* property.
```

>     - *layout*: The dynamic property that holds the layout properties for the chart.

Reviewer: ```suggestion
    - *layout*: The dynamic property that holds the *layout* properties for the chart.
```

>     - *defaultLayout*: The default value for the layout property.

Reviewer: ```suggestion
    - *defaultLayout*: The default value for the *layout* property.
```

> - The **Plot** component utilizes the [`react-plotly.js`](https://github.com/plotly/react-plotly.js) library to render the chart. It accepts two props:

Reviewer: ```suggestion
- The **Plot** component utilizes the [`react-plotly.js`](https://github.com/plotly/react-plotly.js) library to render the chart. It accepts two attributes:
```

    - *data*: An array of data objects to define the chart's datasets.

>     - *layout*: An object specifying the layout properties of the chart.

Reviewer: ```suggestion
    - *layout*: An object specifying the layout of the chart.
```

For more information on how to use `react-plotly.js`, refer to the official documentation [here](https://github.com/plotly/react-plotly.js).

> ## Exporting the React component {data-source="gui:doc/extension/example_library/front-end/src/index.ts"}

Reviewer: ```suggestion
## Exporting the React component {data-source="gui/extension/example_library/front-end/src/index.ts"}
```

When the component is entirely defined, it must be exported by the library's JavaScript bundle.

> This is done by adding the export directive in the file *<project dir>/<package dir>front-end/src/index.ts*.

Reviewer: ```suggestion
This is done by adding the `export` directive in the file *<project dir>/<package dir>front-end/src/index.ts*.
```
is there a missing '/' before "<package_dir>"?

```ts title="index.ts"
import Dashboard from "./Dashboard";

export { Dashboard };
```

> ## Configuring the webpack configuration {data-source="gui:doc/extension/example_library/front-end/webpack.config.js#L39"}

Reviewer: ```suggestion
## Configuring the webpack configuration {data-source="gui/extension/example_library/front-end/webpack.config.js#L39"}
```

To bundle the React component, you must configure the webpack configuration file.

> Since Plotly.js is a JavaScript library, you need to include JavaScript in the webpack’s resolve.extensions setting.

Reviewer: ```suggestion
Since Plotly.js is a JavaScript library, you need to include JavaScript in the webpack’s *resolve.extensions* setting.
```

```js title="webpack.config.js"
module.exports = {
    // ...
    resolve: {
        extensions: [".ts", ".tsx", ".js"],
    },
    // ...
};
```

> After configuring the webpack file, you can build the library as mentioned in the [Building the front-end module](dynamic_element/index.md) section.

Reviewer: Sure about the crossref?

> ## Using the element {data-source="gui:doc/extension/example_library/dashboard.py"}

Reviewer: ```suggestion
## Using the element {data-source="gui/extension/example_library/dashboard.py"}
```

To use the *dashboard* element in a Python script, you can follow the example below:

```python title="dashboard.py"

> trace1 = {

Reviewer: Please rename 'trace' to 'set', to align with the name.
The point is: at the dataset level, we're not talking about traces. This happens only when displayed.
